### PR TITLE
doc(gwk): upload add_algorithm_zh doc

### PR DESCRIPTION
**Pull request by KyeGuo**

### CODE REVIEW

File: docs/how_to_add_new_algorithm_zh.md [hunk 1]
```diff
@@ -1,3 +1,329 @@
...
\ No newline at end of file
+在 Fling 中，我们已经支持一些常见的联邦学习算法（包括通用联邦学习和个性化联邦学习）。考虑到使用者可能需要向框架中添加自定义的联邦学习算法，我们在此作一个简单的介绍，希望能够帮助到您。
+
+## 算法的运行流程：Pipeline（流水线）
+
+无论是通用联邦学习算法还是个性化联邦学习算法，其运行的主体都遵照对应的 pipeline 进行。pipeline 的作用是围绕 Fling 框架中的三个主要组件：客户端（Client）、服务器（Server）和群组（Group）来组织算法的运行流程。主要流程可以分为**配置初始化**和**训练主体**两个部分。
+
+### 配置初始化
+
+这里以个性化联邦学习的 `fling/pipeline/personalized_model_pipeline.py ` 为例，如下所示：
```
Comment (kxzxvbk) [inline docs/how_to_add_new_algorithm_zh.md]: 加上一些串词，如：接下来，我们将一段一段对 pipeline 的主体进行解释。并确保下面的每一段都要和这里呼应上。

File: docs/how_to_add_new_algorithm_zh.md [hunk 2]
```diff
@@ -1,3 +1,329 @@
...
+    )
+    # Logging fine-tune results
+    for key in finetune_results[0][0].keys():
+        for eid in range(len(finetune_results[0])):
+            tmp_mean = sum([finetune_results[cid][eid][key]
+                            for cid in range(len(finetune_results))]) / len(finetune_results)
+            logger.add_scalar(f'finetune/{key}', tmp_mean, eid)
+```
+
+在训练主体部分中，每个通信轮次内，主要涉及**客户端采样**、使用 `launcher` 对各个客户端进行**本地训练**、**聚合前/聚合后测试**，以及使用 `group.aggregate()` 实现服务器组织客户端的**聚合**。由于是个性化联邦学习，在最后还使用 `launcher` 对各个客户端进行本地微调。
```
Comment (kxzxvbk) [inline docs/how_to_add_new_algorithm_zh.md]: 解释一下什么是客户端采样

File: docs/how_to_add_new_algorithm_zh.md [hunk 3]
```diff
@@ -1,3 +1,329 @@
...
+
+**请注意：**此处我们同样建议如非必要，尽量不自定义新的 pipeline，而是选择对相应的组件部分进行修改/添加
```
Comment (kxzxvbk) [inline docs/how_to_add_new_algorithm_zh.md]: 这里有点重复，没必要说两遍尽量不定义新的 pipeline。直接介绍如何对应修改即可

File: docs/how_to_add_new_algorithm_zh.md [hunk 4]
```diff
@@ -1,3 +1,329 @@
...
+
+- 如果涉及到自定义 `logger` 中结果的呈现，可以针对 `train_monitor`、`test_monitor` 以及 `logger.add_scalars_dict()` 部分的操作进行修改。
```
Comment (kxzxvbk) [inline docs/how_to_add_new_algorithm_zh.md:143]: 我觉得这里最关键的点是，要讲明白这里调用的 launch xxx 对应执行的是什么组件的什么函数

File: docs/how_to_add_new_algorithm_zh.md [hunk 5]
```diff
@@ -1,3 +1,329 @@
...
+- 对于**客户端的训练、微调、测试**，**服务器的全局测试**，**群组组织的聚合**和**分发**，可参考[此介绍](https://github.com/KyeGuo/Fling/blob/main/docs/framework_for_fling_zh.md)。具体的自定义方法我们将在下文介绍。
+
+## 自定义算法组件并使用
+
+在详细了解了 pipeline 之后，我们知道要自定义联邦学习算法，可以对客户端（Client）、服务器（Server）和群组（Group）进行自定义，即覆写原组件的方法、添加新属性甚至新方法等。如果需要，甚至可以对流水线（Pipeline）进行自定义。
+
+如果您想向 Fling 添加新的联邦学习算法，可以参考以下步骤：
+
+### 步骤 1：分析需要新定义的组件
```
Comment (kxzxvbk) [inline docs/how_to_add_new_algorithm_zh.md:152]: 这里下面另起一段吧，叫例子之类的

File: docs/how_to_add_new_algorithm_zh.md [hunk 6]
```diff
@@ -1,3 +1,329 @@
...
+        # Reset the global model to save memory.
+        del self.glob_model
+        # Store the current model as prev model in next round
+        self._store_prev_model(self.model)
+        return mean_monitor_variables
+```
+
+在上面的代码示例中，关于新定义组件有几点值得注意：
+
+- 注册新组件：在最开始使用代码 `@CLIENT_REGISTRY.register('fedmoon_client')` 对新定义的 `fedmoon_client` 进行注册，以便在后续的配置文件中使用它。
```
Comment (kxzxvbk) [inline docs/how_to_add_new_algorithm_zh.md:263]: 强调一下，需要在 __init__.py 中 import 这个文件

File: docs/how_to_add_new_algorithm_zh.md (hunk 4 above)
Comment (kxzxvbk) [reply docs/how_to_add_new_algorithm_zh.md:143]: 这一段最为关键，主要讲清楚函数调用的逻辑

File: docs/how_to_add_new_algorithm_zh.md [hunk 7]
```diff
@@ -48,20 +48,21 @@ def personalized_model_pipeline(args: dict, seed: int = 0) -> None:
...
 **请注意：**我们建议如非必要，尽量不自定义新的 pipeline，而是选择对相应的组件部分进行修改/添加。
 
 - 如果要添加数据集，可参考[此教程](https://github.com/KyeGuo/Fling/blob/main/docs/how_to_add_new_dataset_zh.md)。
 - 如果要自定义学习率调度器，可参考[此文档](https://github.com/KyeGuo/Fling/blob/main/docs/meaning_for_configurations_zh.md)中的对应部分。
 - 如果要对客户端、服务器、群组的初始化进行自定义，可参考[此介绍](https://github.com/KyeGuo/Fling/blob/main/docs/framework_for_fling_zh.md)，具体方法我们也会在后文介绍。
+- 启动器 `launcher` 是一个负责组织各个客户端进行训练、测试、微调的组件，具体实现可参考[此处](https://github.com/FLAIR-Community/Fling/blob/main/fling/utils/launcher_utils.py)。
 
 ### 训练主体
 
 这里以个性化联邦学习的 `fling/pipeline/personalized_model_pipeline.py` 为例，如下所示：
```
Comment (kxzxvbk) [inline docs/how_to_add_new_algorithm_zh.md]: 这句话和上面重复了，换一个说法，例如：继续介绍pipeline的xxx部分

File: docs/how_to_add_new_algorithm_zh.md [hunk 8]
```diff
@@ -48,20 +48,21 @@ def personalized_model_pipeline(args: dict, seed: int = 0) -> None:
...
 
 ```python
-# Training loop
+	# Training loop
```
Comment (kxzxvbk) [inline docs/how_to_add_new_algorithm_zh.md]: 去掉这个注释的缩进

File: docs/how_to_add_new_algorithm_zh.md [hunk 9]
```diff
@@ -130,18 +131,19 @@ def personalized_model_pipeline(args: dict, seed: int = 0) -> None:
             logger.add_scalar(f'finetune/{key}', tmp_mean, eid)
 ```
 
-在训练主体部分中，每个通信轮次内，主要涉及**客户端采样**、使用 `launcher` 对各个客户端进行**本地训练**、**聚合前/聚合后测试**，以及使用 `group.aggregate()` 实现服务器组织客户端的**聚合**。由于是个性化联邦学习，在最后还使用 `launcher` 对各个客户端进行本地微调。
+在训练主体部分中，每个通信轮次内，主要涉及**客户端采样**（选取部分客户端参与本轮训练）、使用 `launcher` 对各个客户端进行**本地训练**、**聚合前/聚合后测试**，以及使用 `group.aggregate()` 实现服务器组织客户端的**聚合**。由于是个性化联邦学习，在最后还使用 `launcher` 对各个客户端进行本地微调。
 
-**请注意：**此处我们同样建议如非必要，尽量不自定义新的 pipeline，而是选择对相应的组件部分进行修改/添加
+**请注意：**此处我们同样建议并介绍如何对相应的组件部分进行修改/添加。
```
Comment (kxzxvbk) [inline docs/how_to_add_new_algorithm_zh.md]: 存在语病，修改为：“接下来，我们介绍如何对相应的组件部分进行修改/添加”

File: docs/how_to_add_new_algorithm_zh.md [hunk 10]
```diff
@@ -130,18 +131,19 @@ def personalized_model_pipeline(args: dict, seed: int = 0) -> None:
...
 
+- 使用 `launcher.launch()` 来组织各个客户端串行/并行地执行 `task_name` 对应的操作。具体地，在[代码](https://github.com/FLAIR-Community/Fling/blob/main/fling/utils/launcher_utils.py)中已经定义了对组件 `client` 中的三种内置函数 `train` 、`test` 、`finetune` 的调用。
```
Comment (kxzxvbk) [inline docs/how_to_add_new_algorithm_zh.md]: 说的还是不够清楚，应当类似：1）我们一共有三种可调用的模式，分别是 train test finetune，分别大致的含义是什么；2）真实执行的代码函数在客户端的 class 文件里（可以给出例子说明）；3）我们引入 launcher 的目的是什么（并行化）
